cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,514 @@
+      DeploymentInfo deploymentInfo = new DeploymentInfo(adapterTypeInfo.getFile(), destination);
+      ApplicationWithPrograms applicationWithPrograms = manager.deploy(Id.Namespace.from(namespaceId),
+                                                                       adapterTypeInfo.getType(), deploymentInfo).get();
+      return applicationWithPrograms.getSpecification();
+
+    } catch (Exception e) {
+      throw e;
+    }
+  }
+
+  // Start all the programs needed for the adapter. Currently, only scheduling of workflow is supported.
+  private void startPrograms(String namespaceId, ApplicationSpecification spec, AdapterTypeInfo adapterTypeInfo,
+                             AdapterSpecification adapterSpec) {
+    ProgramType programType = adapterTypeInfo.getProgramType();
+    if (programType.equals(ProgramType.WORKFLOW)) {
+      Map<String, WorkflowSpecification> workflowSpecs = spec.getWorkflows();
+      for (Map.Entry<String, WorkflowSpecification> entry : workflowSpecs.entrySet()) {
+        Id.Program programId = Id.Program.from(namespaceId, spec.getName(), entry.getValue().getName());
+        addSchedule(programId, adapterSpec);
+      }
+    } else {
+      // Only Workflows are supported to be scheduled in the current implementation
+      throw new UnsupportedOperationException(String.format("Unsupported program type %s for adapter",
+                                                            programType.toString()));
+    }
+  }
+
+  // Stop all the programs needed for the adapter. Currently, only unscheduling of workflow is supported.
+  private void stopPrograms(String namespaceId, ApplicationSpecification spec, AdapterTypeInfo adapterTypeInfo,
+                            AdapterSpecification adapterSpec) {
+    ProgramType programType = adapterTypeInfo.getProgramType();
+    if (programType.equals(ProgramType.WORKFLOW)) {
+      Map<String, WorkflowSpecification> workflowSpecs = spec.getWorkflows();
+      for (Map.Entry<String, WorkflowSpecification> entry : workflowSpecs.entrySet()) {
+        Id.Program programId = Id.Program.from(namespaceId, adapterSpec.getType(), entry.getValue().getName());
+        deleteSchedule(programId, SchedulableProgramType.WORKFLOW, adapterSpec.getScheduleName());
+      }
+    } else {
+      // Only Workflows are supported to be unscheduled in the current implementation
+      throw new UnsupportedOperationException(String.format("Unsupported program type %s for adapter",
+                                                            programType.toString()));
+    }
+  }
+
+  // Adds a schedule to the scheduler as well as to the appspec
+  private void addSchedule(Id.Program programId, AdapterSpecification adapterSpec) {
+    String cronExpr = toCronExpr(adapterSpec.getProperties().get("frequency"));
+    Preconditions.checkNotNull(cronExpr, "CronExpression is missing. Cannot schedule program");
```

Comment:
to be more accurate, "frequency" is actually what is missing, not the cron expression